In [ ]:
''' You are an expert in carbon emissions analysis, tasked with estimating the carbon footprint of an individual based on their daily behaviors. 
Your goal is to provide a detailed breakdown of the user's carbon emissions or reductions across various categories of their lifestyle.

Here is the user's description of their daily behaviors:

<user_input>
{{variable}}
</user_input>

Please analyze this input and provide a structured assessment of the user's carbon footprint. Follow these steps:

1. Carefully read the user's input and identify distinct behaviors that may impact their carbon footprint.

2. For each identified behavior:
   a. Categorize it (e.g., Transportation, Home Energy, Gardening, etc.)
   b. Estimate its carbon emission or reduction range per month in kilograms (kg)
   c. Assign a descriptive name to the behavior
   d. Note any additional relevant information provided by the user

3. Compile your analysis into a JSON object following this structure:
   {
     "items": [
       {
         "category": "Category name",
         "emission_range": "X-Y kg",
         "name": "Descriptive name of the behavior",
         "additional_info": "Any extra details provided"
       },
       // Additional items as needed
     ]
   }

Note:
- Emission ranges should be given as "X-Y kg" where X and Y are numbers.
- Negative emission ranges (e.g., "-5--1 kg") indicate carbon reduction activities.
- Be as precise as possible with your estimates, but understand that these are approximations.

Before providing your final output, wrap your analysis in <behavior_analysis> tags. In this analysis:

1. List all identified behaviors, numbering them for easy reference.
2. For each behavior:
   a. Quote the relevant part of the user input
   b. Explain your reasoning for the category assignment
   c. Show your calculation for the emission estimate, breaking it down step-by-step
   d. Justify the emission range based on potential variations
   e. Explain your choice of the descriptive name

Be thorough in your analysis; it's okay for this section to be quite long. This will help ensure a comprehensive and accurate assessment.

Example analysis structure (do not copy this content, it's just for format reference):
<behavior_analysis>
Identified behaviors:
1. Daily car commute
2. [Other behaviors...]

Behavior 1: Daily car commute
User input: "[Quote relevant part of user input]"
Category: Transportation
Reasoning for category: This behavior involves personal vehicle use for regular travel, which falls under the transportation category.
Emission calculation:
- Assumed distance: 30 miles round trip
- Vehicle: 2013 Honda Accord (typical emission: 0.35 kg CO2 per mile)
- Daily emission: 30 miles * 0.35 kg/mile = 10.5 kg
- Monthly emission (22 working days): 10.5 kg * 22 = 231 kg
Emission range justification: Accounting for potential variations in traffic conditions and exact working days, a range of 220-240 kg per month is reasonable.
Name explanation: "Daily car commute" clearly describes the regular nature and purpose of this transportation behavior.
Additional info: 2013 Honda Accord, 30 miles round trip

[Continue with other behaviors...]
</behavior_analysis>

Once you've completed your analysis, provide the final JSON output as described above. '''

In [22]:
import openai

#latest version

In [21]:
openai.api_key = ""
temperature_set = 0.0
model_name = "gpt-3.5-turbo-0125"  # "gpt-3.5-turbo-0125" #"gpt-3.5-turbo-16k"# "gpt-4"

In [12]:
t1 = '''You are an expert in carbon emissions analysis, tasked with estimating the carbon footprint of an individual based on their daily behaviors. Your goal is to provide a detailed breakdown of the user's carbon emissions or reductions across various categories of their lifestyle.

Here is the user's description of their daily behaviors:'''


t2 = '''Please analyze this input and provide a structured assessment of the user's carbon footprint. Follow these steps:

1. Carefully read the user's input and identify distinct behaviors that may impact their carbon footprint.

2. For each identified behavior:
   a. Categorize it (e.g., Transportation, Home Energy, Gardening, etc.)
   b. Estimate its carbon emission or reduction range per month in kilograms (kg)
   c. Assign a descriptive name to the behavior
   d. Note any additional relevant information provided by the user

3. Compile your analysis into a JSON object following this structure:
   {
     "items": [
       {
         "category": "Category name",
         "emission_range": "X-Y kg",
         "name": "Descriptive name of the behavior",
         "additional_info": "Any extra details provided"
       },
       // Additional items as needed
     ]
   }

4. If the user is providing informationa bout members of his family or multiple people in his home, then provide ALL the emission ranges cumulatively for all members. 
Stay consistent in your emission estimates - of whether you are doing it for a single person or multiple people. 

4. Based on the emission values and ranges, provide a final assessment of the user (or the total of the user's family/multiple members depending on the case) carbon footprint in the format: \" Estimated monthly carbon footprint: \"x\" kg \" .
It should be a single numerical value in kilograms (kg) representing the estimated monthly carbon emissions. Strictly stick to the format mentioned above for the final estimate.
If you are providing a range, please provide the average of the range as the final estimate.

Note:
- Emission ranges except the final estimate should be given as "X - Y kg" where X and Y are numbers.
- Negative emission ranges (e.g., "-5 - -1 kg") indicate carbon reduction activities.
- Be as precise as possible with your estimates, but understand that these are approximations.


Provide the final JSON output as described above with the estimated monthly carbon emissions. '''

In [13]:
user_input = ''' I drive a 2013 Honda accord 15 miles to and from my work to my house. I have a family of 4 with my wife and 2 kids, all of us are non-vegetarians. We have a double door refrigerator and we live in Atlanta in a 2b2b appartment. I keep my thermostat around 68° fahrenheit in summers and 76 degrees fahrenheit in winters . We use all basic appliances like dishwasher, toaster, HVAC, and we have 4 laptops, phones and other small gadgets. In my small balcony which gets ample sunlight, I have 3-4 potted plants.  All the bulbs in my house are LED bulbs. '''

In [3]:
prompt = t1 + '\n\n ##### \n' \
    + user_input + "\n\n" \
    + t2

In [16]:
# Create OpenAI client and get response
client = openai.Client(api_key=openai.api_key)

response = client.chat.completions.create(
            model=model_name,
            messages=[{"role": "user", "content": prompt}],
            temperature=temperature_set,
            max_tokens=2000,
            )

In [17]:
response.choices[0].message.content

'{\n  "items": [\n    {\n      "category": "Transportation",\n      "emission_range": "80 - 100 kg",\n      "name": "Daily commute in a 2013 Honda Accord",\n      "additional_info": "15 miles round trip to work"\n    },\n    {\n      "category": "Food",\n      "emission_range": "150 - 200 kg",\n      "name": "Non-vegetarian diet for a family of 4",\n      "additional_info": "Family of 4 including 2 kids"\n    },\n    {\n      "category": "Home Energy",\n      "emission_range": "50 - 70 kg",\n      "name": "Home energy usage",\n      "additional_info": "2b2b apartment in Atlanta, LED bulbs, basic appliances, thermostat set at 68°F in summers and 76°F in winters"\n    },\n    {\n      "category": "Gardening",\n      "emission_range": "-5 - -1 kg",\n      "name": "Potted plants on balcony",\n      "additional_info": "3-4 potted plants receiving ample sunlight"\n    }\n  ]\n} \n\nEstimated monthly carbon footprint: 275 kg'

### Response for the same input- 

1. '{\n  "items": [\n    {\n      "category": "Transportation",\n      "emission_range": "80 - 100 kg",\n      "name": "Daily commute in a 2013 Honda Accord",\n      "additional_info": "15 miles round trip to work"\n    },\n    {\n      "category": "Home Energy",\n      "emission_range": "50 - 70 kg",\n      "name": "Home energy usage",\n      "additional_info": "Maintaining thermostat at 68°F in summers and 76°F in winters, basic appliances, LED bulbs"\n    },\n    {\n      "category": "Gardening",\n      "emission_range": "-5 - -1 kg",\n      "name": "Potted plants on balcony",\n      "additional_info": "3-4 potted plants receiving ample sunlight"\n    }\n  ]\n}\n\nEstimated monthly carbon footprint: 75 kg'

2. '{\n  "items": [\n    {\n      "category": "Transportation",\n      "emission_range": "80-100 kg",\n      "name": "Daily commute in a 2013 Honda Accord",\n      "additional_info": "15 miles round trip to work"\n    },\n    {\n      "category": "Food",\n      "emission_range": "150-200 kg",\n      "name": "Non-vegetarian diet for a family of 4",\n      "additional_info": "Family of 4 consuming meat"\n    },\n    {\n      "category": "Home Energy",\n      "emission_range": "50-70 kg",\n      "name": "Home energy usage",\n      "additional_info": "Thermostat set at 68°F in summers and 76°F in winters, basic appliances, LED bulbs"\n    },\n    {\n      "category": "Gardening",\n      "emission_range": "-5 - -1 kg",\n      "name": "Potted plants on balcony",\n      "additional_info": "3-4 potted plants receiving ample sunlight"\n    }\n  ]\n} \n\nEstimated monthly carbon footprint: 275 kg'

3. '{\n  "items": [\n    {\n      "category": "Transportation",\n      "emission_range": "80 - 100 kg",\n      "name": "Daily commute in a 2013 Honda Accord",\n      "additional_info": "15 miles round trip to work"\n    },\n    {\n      "category": "Food",\n      "emission_range": "150 - 200 kg",\n      "name": "Non-vegetarian diet for a family of 4",\n      "additional_info": "Family of 4 consuming meat"\n    },\n    {\n      "category": "Home Energy",\n      "emission_range": "50 - 70 kg",\n      "name": "Home energy usage",\n      "additional_info": "Maintaining thermostat at 68°F in summers and 76°F in winters, using basic appliances"\n    },\n    {\n      "category": "Home Energy",\n      "emission_range": "5 - 10 kg",\n      "name": "LED bulbs usage",\n      "additional_info": "All bulbs in the house are LED"\n    },\n    {\n      "category": "Gardening",\n      "emission_range": "-2 - 3 kg",\n      "name": "Potted plants on balcony",\n      "additional_info": "3-4 potted plants receiving ample sunlight"\n    }\n  ]\n}\n\nEstimated monthly carbon footprint: 283 kg'

4. '{\n  "items": [\n    {\n      "category": "Transportation",\n      "emission_range": "80 - 100 kg",\n      "name": "Daily commute in a 2013 Honda Accord",\n      "additional_info": "15 miles round trip to work"\n    },\n    {\n      "category": "Food",\n      "emission_range": "150 - 200 kg",\n      "name": "Non-vegetarian diet for a family of 4",\n      "additional_info": "Family of 4 including 2 kids"\n    },\n    {\n      "category": "Home Energy",\n      "emission_range": "50 - 70 kg",\n      "name": "Home energy usage",\n      "additional_info": "2b2b apartment in Atlanta, LED bulbs, basic appliances, thermostat set at 68°F in summers and 76°F in winters"\n    },\n    {\n      "category": "Gardening",\n      "emission_range": "-5 - -1 kg",\n      "name": "Potted plants on balcony",\n      "additional_info": "3-4 potted plants receiving ample sunlight"\n    }\n  ]\n} \n\nEstimated monthly carbon footprint: 275 kg'

### Phase 1
- Get initial free text user input
- Call to LLM to get estimates
- Show the montly estimate at the top to the user
- Show the analysis as well with a "view details" option
- Ask the user for :budget and category in which they'd like to make changes
- Call to LLM to get recommendations
- Call to db to get suggestions of local vendors and nurseries for plants and shops for elctrical appliances ( will set us apart from shadow-LLM-based app)

### Phase 2
- Fine-tune LLM with datasets ( found) to improve predictions
- Iterative refinement capabilities

In [19]:
t3 = '''You are an environmental consultant tasked with providing personalized eco-friendly recommendations. Given the following user data and preferences, provide specific, actionable recommendations organized by category.

Current Environmental Impact:
{extracted user data JSON}

Monthly Carbon Footprint: 275 kg

User Preferences:
- Monthly Budget: ${user's input}
- Selected Categories: {list of selected categories}

For each selected category, provide 2-3 specific recommendations following this structure:
- Recommendation title
- Implementation steps
- Estimated cost
- Expected carbon reduction
- Additional benefits
- Local resources needed (if any)
- Timeline for implementation
- Maintenance requirements

Guidelines for recommendations:
1. Start with the easiest/highest-impact-per-dollar options first
2. Consider the user's current lifestyle (based on their data)
3. Account for their location (Atlanta) and living situation (apartment)
4. Ensure recommendations are within the specified monthly budget
5. Focus on measurable impact
6. Consider seasonal factors
7. Include specific product types where relevant (but no brand names)
8. Consider the family size (4 members) when suggesting modifications

Response format example:

Category: {Category Name}
1. {Recommendation Title}
   - Implementation: {detailed steps}
   - Cost: {specific amount}
   - Carbon Reduction: {kg/month}
   - Benefits: {list specific benefits}
   - Resources: {what's needed}
   - Timeline: {specific timeframe}
   - Maintenance: {requirements}

[Repeat for each recommendation]

Please provide practical, specific recommendations that acknowledge their current lifestyle while suggesting meaningful improvements.'''